### PR TITLE
🎉 Source Salesforce: add optional parameters to filter properties within streams

### PR DESCRIPTION
**Pull request by jkaelin**
## What
Allows a user to exclude specific fields and/or data-types from the stream.

### Current Issues
There are at least a few issues directly requesting the generic ability to exclude specific fields from streams:
- [Allow a user to filter fields/columns from a table #2227](https://github.com/airbytehq/airbyte/issues/2227)
- [Allow excluding volatile fields in standard tests #1675](https://github.com/airbytehq/airbyte/issues/1675)

I acknowledge that this PR is specific to the Salesforce Source and does not fully address the above issues.

### General Use Cases
Some use cases that may be faciliated by this PR:
- Exclude sensitive fields or data types from streams in order to comply with legal/privacy regulations
- Exclude fields or data types which may prohibit the stream from syncing successfully
- Exclude datatypes which may invoke the fallback to the non-Bulk API (e.g. `base64`/`object`)

### Specific Use Case
The immediate use case being solved for with this implementation is the ability to exclude certain calculated fields within Salesforce objects that may generate exceptions on evaluation. Although these exceptions are effectively masked in the UI, they will cause the Salesforce Bulk API to fail with only a generic error message.

## How
- Add two optional string array parameters to the spec
- Inject evaluation of these user exclusions against the fields returned by `generate_schema`

## Recommended reading order
1. `spec.json`
    - Added two optional parameters `exclude_fields` & `exclude_types`
2. `source.py`
    - `get_user_excluded_fields` & `get_user_excluded_types`
      - Added two helper functions to evaluate the above parameters
    - `generate_streams`
      - Pass parameters on to `api.py`:`generate_schema`
4. `api.py`
    - `generate_schema`
      - add optional parameters for user exclusions
      - exclude fields from schema based on optional parameters
4. `unit_test.py`
    - `test_bulk_sync_pagination`
      - No functional changes
      - Forced to add the `#noqa E203` directive due to conflict between `black` & `flake8` on line 143
    - `test_schema_with_user_excluded_fields`
      - parameterized unit test for named fields
    - `test_schema_with_user_excluded_types`
      - parameterized unit test for data types

## User Impact

- Users will see two new optional parameters on the New Source page
  - Fields to Exclude
    msedge_EY1gpJuSHa  
  - Data Types to Exclude
    msedge_zwrrmF7FuN  
- Sync logging will explicitly list the fields which are being skipped

## Pre-merge Checklist
<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests)) -- will do
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

### CODE REVIEW

**Comment by harshithmullapudi** [comment]
Hey, thanks for the contribution can you sourceAcceptanceTests screenshot?

**Comment by jkaelin** [comment]
Hello, is this what you need?

[image: image.png]

*Few questions for you*

The pre-commit hooks often formats/flags code that I didn't write.
Is it correct behavior to submit these changes with the PR?
I had to use the noqa directive due to conflicting black/flake behavior. Is
that the correct way to resolve this type of issue?

I made a mistake in this PR - should I close the PR and re-submit?
I realize that a new PR would cause some churn on your side since there are
already some processes underway.
Please advise if there is a preference to how this situation is handled.

Thank you!

**Comment by jkaelin** [comment]
@harshithmullapudi 
Is this what you need for the _sourceAcceptanceTests_?
image

**Comment by jkaelin** [comment]
**Question for maintainers**
The pre-commit hooks often formats/flags code that I didn't write.
Is it correct behavior to submit these changes with the PR?
I had to use the noqa directive due to conflicting black/flake behavior. 
Is that the correct way to resolve this type of issue?

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py [hunk 1]
```diff
@@ -35,6 +35,25 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
                 logger.warn(f"API Call limit is exceeded. Error message: '{error_data.get('message')}'")
                 return False, "API Call limit is exceeded"
 
+    @staticmethod
+    def get_user_excluded_fields(config: Mapping[str, Any], stream_name: str) -> List[str]:
+        """Retrieve user-excluded fields (specified as Stream.Field) for this stream from config"""
+        excluded_fields = []
+        if config.get("exclude_fields") and stream_name is not None:
+            for f in config["exclude_fields"]:
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py:43]:
```suggestion
            for field_name in config["exclude_fields"]:
```

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py [hunk 2]
```diff
@@ -35,6 +35,25 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
...
+                if "." in f:
+                    if f.split(".")[0] == stream_name:
+                        excluded_fields.append(f.split(".")[1])
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py:46]:
```suggestion
        if config.get("exclude_fields") and stream_name is not None:
            for stream_field in config["exclude_fields"]:
                 try:
                     exclude_stream_name, exclude_field_name = stream_field.split(".")
                 exception ValueError as e:
                     raise "Stream.Field specified is not correct"
                 if exclude_stream_name == stream_name:
                      excluded_fields.append(exclude_field_name)
                      
```

What do you think about this?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py (hunk 2 above)
Comment (jkaelin) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py:46]:
```suggestion
        if config.get("exclude_fields") and stream_name is not None:
            for stream_field in config["exclude_fields"]:
                 try:
                     exclude_stream_name, exclude_field_name = stream_field.split(".")
                 exception ValueError as e:
                     raise f"Format for excluded field [{stream_field}] is incorrect. Expected format is Stream.Field"
                 if exclude_stream_name == stream_name:
                      excluded_fields.append(exclude_field_name)
```
Variable name change makes sense.
I think your recommendation of the split into two variables is easier to read.
try...catch makes sense but if we are going to raise an exception on formatting error, I think we should include the input details in addition to the error description to make it easier for the user to resolve.

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json [hunk 3]
```diff
@@ -60,24 +66,49 @@
...
+        "order": 110
+      },
+      "exclude_fields": {
+        "title": "Fields to Exclude",
+        "description": "Add any fields that should always be excluded using Stream.Field syntax",
+        "examples": ["Contact.SSN", "CustomerPayment.PrimaryCC"],
+        "type": "array",
+        "items": {
+          "type": "string",
+          "pattern": "^[A-Za-z0-9_]+\\.[A-Za-z0-9_]+$"
```
Comment (jkaelin) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json:90]: @marcosmarxm updated the spec here to enforce the Stream.Field syntax - would be best to catch these issues up front!

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py (hunk 2 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py:46]: Of course! much better.

**Comment by marcosmarxm** [comment]
There is some conflicts because other code was merged, can you solved it @jkaelin ?

**Comment by jkaelin** [comment]
@marcosmarxm made the changes you requested & rebased to latest.

**Comment by jkaelin** [comment]
@marcosmarxm anything else you need me to address here?

**Comment by marcosmarxm** [comment]
I'll test your code tomorrow @jkaelin !

**Review by alafanechere — APPROVED** [review]
Thanks for this contribution! I slightly changed the custom exception you tried to raise, which broke the test.
@misteryeo / @sherifnada, please give this a quick review as the `spec.json` was updated: two new optional fields.

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py [hunk 4]
```diff
@@ -35,13 +35,34 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
                 logger.warn(f"API Call limit is exceeded. Error message: '{error_data.get('message')}'")
                 return False, "API Call limit is exceeded"
 
+    @staticmethod
+    def get_user_excluded_fields(config: Mapping[str, Any], stream_name: str) -> List[str]:
+        excluded_fields = []
+        if config.get("exclude_fields") and stream_name is not None:
+            for stream_field in config["exclude_fields"]:
```
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py]: I made the change in https://github.com/airbytehq/airbyte/pull/9931/commits/8092db2c12ccb160b296aca477dd6f7df538a710

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py [hunk 5]
```diff
@@ -35,13 +35,34 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
...
+                try:
+                    exclude_stream_name, exclude_field_name = stream_field.split(".")
+                except ValueError:
+                    raise f"Format for excluded field [{stream_field}] is incorrect. Expected format is Stream.Field"
```
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py]: I made the change in https://github.com/airbytehq/airbyte/pull/9931/commits/8092db2c12ccb160b296aca477dd6f7df538a710

**Review by misteryeo — APPROVED** [review]
LGTM!

**Comment by misteryeo** [comment]
Don't think this is blocking but I'm curious why the implementation here is just a field and not the variable input that we currently have for "Streams Filter Criteria" https://share.getcloudapp.com/P8u6njow

**Comment by misteryeo** [comment]
Thanks, appreciate the insight @jkaelin - we'll definitely look to improve the way the variable input module works. cc: @timroes

**Comment by alafanechere** [comment]
Hey @jkaelin an other PR was merged on this connector, leading to multiple conflicts, do you mind fixing these?

**Comment by alafanechere** [comment]
Hi @jkaelin, we closed this PR as we did not receive updates following our latest interactions. You are welcome to re-open it as soon as you want. We’ll be happy to give additional reviews. Do not hesitate to reach our team on Slack if you need support finishing this work.

**Pull request closed without merging**